Update to `vinyl@2.0.0`

demurgos opened the pull request:
Update major version of `vinyl`:
- **Breaking change**: remove `PipeOptions` interface
- **Breaking change**: remove `pipe` method
- Update documentation
- Update source-tests

This PR still contains the undocumented `stat` property.

As @felixfbecker said it in https://github.com/types/npm-vinyl/pull/2#issuecomment-254422414, please create a branch for `1.x` before updating the typings for `vinyl@2.0.0`

felixfbecker commented on source-test/tsconfig.json:
```diff
@@ -5,7 +5,9 @@
   },
   "files": [
-    "File.ts",
+    "file.ts",
```
Is there a reason you renamed this file? git sees it as a delete+add rather than a modify

felixfbecker replied:
but isn't this just the tests? The tests can be named whatever we want

felixfbecker commented on source-test/tsconfig.json:
```diff
@@ -5,7 +5,9 @@
-    "File.ts",
+    "file.ts",
+    "inspect-stream.ts",
+    "normalize",
```
Missing .ts

demurgos commented:
I just restored the original capitalization of `File.ts` and fixed the reference to `normalize.ts` in `source-test/tsconfig.json`

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
-import File = require('vinyl');
-import * as Stream from 'stream';
-import * as fs from 'fs';
+var fs = require('fs');
```
Use `import * as fs from 'fs'`, and where not possible, use `import x = require('x');`

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
+var concat = miss.concat;
+var isCloneable = cloneable.isCloneable;
+
+var isWin = (process.platform === 'win32');
```
Unneeded parens

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
-      const file = new File({ cwd: cwd });
-      file.basename.should.equal(cwd);
+    it('defaults cwd to process.cwd', function(done) {
+      var file = new Vinyl();
```
Use `const`

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
+
+      pipe([
+        file.contents,
+        concat(function(d: Buffer) {
```
arrow function

demurgos commented:
I did my best to convert the tests to ES6, could you check this ? I just hope that integrating upstream changes from `vinyl` will not require to convert them again every time.

During the conversion, I encountered the following issues:
- There is no definition for `mississippi`
- There is no definition for `cloneable-readable`
- @andrew-w-ross https://github.com/andrew-w-ross/typings-expect/issues/2: Missing aliases in the definitions  of `expect`
- @andrew-w-ross https://github.com/andrew-w-ross/typings-expect/issues/3: Implicity `any`s in the definitions of `expect`
- https://github.com/types/env-node/issues/42: `fs.Stats` should be declared as a class.

To solve them, I added placeholders for modules, deactivated `noImplicitAny` for the tests (we still keep it for the declarations) and had to rename and cast some values.

blakeembrey commented:
If @andrew-w-ross doesn't respond, we can look at forking the typings into @types. I know he's been busy offline before.

felixfbecker commented:
@demurgos There is a _lot_ of casting going on which gives me the feeling that we are actually not testing anything here because we cast so much to `any`. We need to fix these issues. A dependent library with outdated defs is okay but casting input to vinyl functions is not. It is also okay to simply require the libraries we don't have defs for.

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
-      file.basename.should.equal(cwd);
+    it('sets custom properties', function(done) {
+      const sourceMap = {};
+      const file = <Vinyl & {sourceMap: {}}> new Vinyl(<any> { sourceMap: sourceMap });
```
For example this: If we need to support custom properties, we should add `[attribute: string]: any` to the options signature instead of requiring a cast to `any`.

demurgos replied:
I was wondering if it was possible to tell typescript that the class implements its options. Here is a solution I tried but it did not seem to work:

```typescript
interface FileOptions {
  src: string;
}
// This does not work
class File<O extends {}> implements O {
  constructor (options?: FileOptions & O);
}

// This does not work either
class File<O extends {}> {
  constructor (options?: FileOptions & O): this & O;
}
```

The best I could come up with is the solution I used in the tests where the user augments the object:

```typescript
interface Custom {
  custom: number;
}

const file = <Vinyl & Custom> new Vinyl(<FileOptions & Custom> {src: "...", custom: 10});
// It requires more work from the user, but why not. The problem is `clone`:
const file2 = file.clone();  // file2 is typed as Vinyl, not Vinyl & Custom (you have to cast it manually)
```

Do you have some ideas ?

(Of course, I'll add the string property accessor on options, but how to express the dependency between the options and the properties of the object ?)

felixfbecker replied:
That's not possible atm, casting to a union type is the only way.

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
-      const file = new File(options);
-      const file2 = file.clone();
+  describe('isDirectory()', function() {
+    const fakeStat: fs.Stats = <any> {
```
The same here:
 - if vinyl allows to pass just an object with `isDirectory()`, the typings should allow to pass just an object with `isDirectory()`.
 - if vinyl requires a full-fledged `Stat` object, the tests should pass a full-fledged `Stat` object.

demurgos replied:
I'll contact the maintainer of Vinyl again. It is documented as requiring a full-fledged `Stats` but in fact it only uses two methods. I think that this is because he plans to support full-fledged `Stats` (https://github.com/gulpjs/vinyl/issues/105) in the near future so requiring it right now can facilitate the transition.

felixfbecker replied:
Then just do `new Stat()` here. `@phated` is a very busy person ;)

demurgos replied:
Node's `Stats` constructor is not even documented. I could use it ([see the source](https://github.com/nodejs/node/blob/master/lib/fs.js#L132)) but which values should I pass it ?
Vinyl tests a real `Stats` object [only once](https://github.com/gulpjs/vinyl/blob/master/test/file.js#L601).

Since Vinyl is supposed to handle "virtual" files, I kinda understand that its tests do not completely match what is documented: it relies on knowledge of which internal methods are called. If you look trough the code, it sometimes defines only `isDirectory`, sometimes `isSymbolicLink`, sometimes both. I agree that it would be better if he only used real `Stats` object since it is documented as this, but do we really need to change it if we are never going to run the tests ?

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
+      expect(file2.base).toEqual(file.base);
+      expect(file2.path).toEqual(file.path);
+      expect(file2.contents).toNotBe(file.contents);
+      expect((<Buffer> file2.contents).toString('utf8')).toEqual((<Buffer> file.contents).toString('utf8'));
```
Why do we need to cast here? If `file.contents` is a `Buffer`, it must be typed as one. If we cast it we don't test anything here, as we cannot expect the user to cast. If it may be something different, we assert that it is one first.

demurgos replied:
[It can be a readable stream.](https://github.com/gulpjs/vinyl#filecontents)

felixfbecker replied:
Then first assert that it is a `Buffer`

demurgos replied:
Which style do you prefer ?

This one is the closer to the original tests:
```typescript
expect(file2.contents).toBeA(Buffer);
expect(file.contents).toBeA(Buffer);
expect((file2.contents as Buffer).toString('utf8')).toEqual((file.contents as Buffer).toString('utf8'));
```
Or this one that Typescript understands (so no need to manually add type assertion):
```typescript
if (!(file2.contents instanceof Buffer)) {
  throw new TypeError("Expected `file2.contents` to be a `Buffer`");
}
if (!(file.contents instanceof Buffer)) {
  throw new TypeError("Expected `file.contents` to be a `Buffer`");
}
expect(file2.contents.toString('utf8')).toEqual(file.contents.toString('utf8'));
```

In fact, I'm fearing that it is starting to deviate a bit too much from the original tests: we are not really running the tests, just compiling them. The fact that the cast to `Buffer` is allowed is already good enough (for example, it wouldn't let you cast to `Array`) to check the definitions.

felixfbecker replied:
The first

demurgos replied:
Ok, I preferred it too because it is closer to the source.

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
         base: '/test/',
         path: '/test/test.coffee',
-        contents: fakeStream
+        contents: <Buffer> null,
```
If `contents` accepts `null`, the typings should allow `null`. If it doesn't, then the tests shouldn't pass `null`.

demurgos replied:
I guess it's related to the above: the compiler does not know if it is a Buffer or a ReadableSteam.

felixfbecker replied:
`null` is neither one in TS2, `null` is `null`

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
+      // Just copy static stuff since Object.setPrototypeOf is node >=0.12
+      Object.keys(Vinyl).forEach((key) => {
+        (<any> ExtendedFile)[key] = (<{[key: string]: any}> Vinyl)[key];
       });
```
Since this is TypeScript that gets compiled, you can test this with `class ExtendedFile extends Vinyl {}`

demurgos replied:
Ok, I'll use the class syntax.

felixfbecker commented on source-test/File.ts:
```diff
@@ -1,661 +1,1617 @@
+
+    it('returns _contents', function(done) {
+      const val = new Buffer('test');
+      const file: Vinyl & {_contents?: Buffer} = new Vinyl();
```
👍 This is what I'm looking for - showing how to accomplish stuff without casting

demurgos replied:
Ok, maybe I could it in a section in the README.md ?

felixfbecker replied:
If you want. Not required

demurgos commented:
Ok, thank you for you various comments, I'll work on this in the next days.
Some of the casting is caused by the fact that it uses properties that are not part of public API (`_contents` for example) but I'll try to reduce the amount of casts.

felixfbecker commented:
@demurgos Thank you for all your work, really looking forward to this.

demurgos commented:
@blakeembrey 
> If @andrew-w-ross doesn't respond, we can look at forking the typings into @types. I know he's been busy offline before. 

Ok, I'll finish the update of this module and then send a PR for `expect`. I'll also check if I can update the typings I wrote in may (like "big-integer" or "npmlog") now that I could gain a bit more experience. Could I move them to `@types` once they met the `typings` quality ? (I never moved repositories between organizations so I'm not sure about how it works)

andrew-w-ross commented:
@blakeembrey I can move all of my type definitions over to the types organization if you want.

blakeembrey commented:
Thanks @andrew-w-ross, I'll take a look and make some updates. GitHub will handle the redirects until then.

felixfbecker commented on index.d.ts:
```diff
@@ -23,7 +23,7 @@ declare namespace File {
      * Default: `undefined`
      */
-    path?: string;
+    path?: string | null | undefined;
```
Afaik `?` is the same as `| undefined`

demurgos replied:
Ok, I did not remember the rule for interfaces (for functions `?` is `| undefined` but I wasn't sure). I'll update it.

demurgos commented:
Hi,
I added strict null checks, replace `Stats`-like objects by `Stats` instances, used a TS class extension for the test of the inheritance, updated the dependencies, and added support for custom properties.

demurgos commented on source-test/File.ts:
```diff
@@ -5,6 +5,10 @@ import expect = require('expect');
 import cloneable = require('cloneable-readable');
 
+interface ObjectConstructorWithAssign extends ObjectConstructor {
+  assign<T>(target: T, ...sources: {}[]): T
```
This should, but I did not have it (TS compiler 2.0.3). Maybe in the lib ES6 ?

felixfbecker replied:
Probably needs `target: ES6` for the tests 😉

demurgos commented:
Hi, I removed the unnecessary `| undefined` and changed the target of the tests to `es6`. This solved the missing definition for `Object.assign`.

demurgos commented:
Well, I just remembered this PR. I'll check if everything still works and merge it if it works.

demurgos commented:
I rebased onto the latest version of master. I'll check if the typings are still correct.

demurgos commented:
It seems that the branch was already ready. I just did some minor updates to ensure that the installation and tests run fine.

@felixfbecker, @blakeembrey
I know that you already heavily reviewed this PR. If you want to take a look at it, I'll leave it open this week, otherwise I'll simply merge it.

blakeembrey commented:
@demurgos I think that's ok, it looks like the changes were previously accepted. Just remember to branch the current commit on `master` into a `1.x` branch when you land this.
